# Background

In the train.csv (and test.csv) data, features that belong to similar groupings are tagged as such in the feature names (e.g., ind, reg, car, calc). In addition, feature names include the postfix bin to indicate binary features and cat to indicate categorical features. Features without these designations are either continuous or ordinal. Values of -1 indicate that the feature was missing from the observation. The target columns signifies whether or not a claim was filed for that policy holder.

In [142]:
import pandas as pd
import numpy as np

# Read data

In [143]:
df = pd.read_csv('data/train.csv',index_col=False)

In [144]:
df.shape

(595212, 59)

In [5]:
df.head()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0   7       0          2              2          5              1   
1   9       0          1              1          7              0   
2  13       0          5              4          9              1   
3  16       0          0              1          2              0   
4  17       0          0              2          0              1   

   ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin       ...        \
0              0              0              1              0       ...         
1              0              0              0              1       ...         
2              0              0              0              1       ...         
3              0              1              0              0       ...         
4              0              1              0              0       ...         

   ps_calc_11  ps_calc_12  ps_calc_13  ps_calc_14  ps_calc_15_bin  \
0           9           1           5           8               0   
1           3           1           1           9               0   
2           4           2           7           7               0   
3           2           2           4           9               0   
4           3           1           1           3               0   

   ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  ps_calc_19_bin  \
0               1               1               0               0   
1               1               1               0               1   
2               1               1               0               1   
3               0               0               0               0   
4               0               0               1               1   

   ps_calc_20_bin  
0               1  
1               0  
2               0  
3               0  
4               0  

[5 rows x 59 columns]

In [6]:
df.loc[df.target == 1].head()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
9    28       1          1              1          2              0   
19   64       1          0              1          2              1   
28   84       1          0              2          0              1   
39  107       1          2              1          3              1   
41  110       1          5              2          1              0   

    ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
9               0              0              1              0   
19              0              1              0              0   
28              4              1              0              0   
39              0              0              0              1   
41              0              0              0              1   

         ...        ps_calc_11  ps_calc_12  ps_calc_13  ps_calc_14  \
9        ...                 3           5           0           6   
19       ...                10           3           1          11   
28       ...                 3           2           0           8   
39       ...                 6           0           8          14   
41       ...                 5           2           0           7   

    ps_calc_15_bin  ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  \
9                0               1               0               0   
19               0               1               1               0   
28               0               1               1               0   
39               0               1               0               1   
41               1               1               0               0   

    ps_calc_19_bin  ps_calc_20_bin  
9                1               0  
19               1               0  
28               0               0  
39               0               0  
41               0               0  

[5 rows x 59 columns]

## Q1: How many explanatory features are there in the training data (train.csv)? Could id be considered as an explanatory feature and why? 

### Answer 
1. There are 57 explanatory features here, if we do not include the label -- 'target'. 

2. And 'id' cannot be considered as an explanatory feature in this case, since it is only the index column for the dataset. There is no meaning for this column.

Below are codes to explain the answer above.

In [ ]:
df.info()

## Q2: Among the explanatory features in train.csv, how many are nominal (i.e., binary or categorical)? Please list all nominal features respectively. For each of them, please provide the count of each category.

### Answer 
1. There are 31 nominal data (binary and categorical), as can be shown in col_Q2 list.

2. For each of them, the count of each category is in Q2.

In [7]:
col_Q2 = df.columns[[v[-3:] in 'cat and bin' for v in df.columns]]

In [8]:
col_Q2

Index(['ps_ind_02_cat', 'ps_ind_04_cat', 'ps_ind_05_cat', 'ps_ind_06_bin',
       'ps_ind_07_bin', 'ps_ind_08_bin', 'ps_ind_09_bin', 'ps_ind_10_bin',
       'ps_ind_11_bin', 'ps_ind_12_bin', 'ps_ind_13_bin', 'ps_ind_16_bin',
       'ps_ind_17_bin', 'ps_ind_18_bin', 'ps_car_01_cat', 'ps_car_02_cat',
       'ps_car_03_cat', 'ps_car_04_cat', 'ps_car_05_cat', 'ps_car_06_cat',
       'ps_car_07_cat', 'ps_car_08_cat', 'ps_car_09_cat', 'ps_car_10_cat',
       'ps_car_11_cat', 'ps_calc_15_bin', 'ps_calc_16_bin', 'ps_calc_17_bin',
       'ps_calc_18_bin', 'ps_calc_19_bin', 'ps_calc_20_bin'],
      dtype='object')

In [249]:
col_Q2.shape

(31,)

In [9]:
Q2 = [df[col].value_counts() for col in df[col_Q2]]

In [10]:
Q2

[ 1    431859
  2    123573
  3     28186
  4     11378
 -1       216
 Name: ps_ind_02_cat, dtype: int64,  0    346965
  1    248164
 -1        83
 Name: ps_ind_04_cat, dtype: int64,  0    528009
  6     20662
  4     18344
  1      8322
  3      8233
 -1      5809
  2      4184
  5      1649
 Name: ps_ind_05_cat, dtype: int64, 0    360852
 1    234360
 Name: ps_ind_06_bin, dtype: int64, 0    442223
 1    152989
 Name: ps_ind_07_bin, dtype: int64, 0    497644
 1     97568
 Name: ps_ind_08_bin, dtype: int64, 0    484917
 1    110295
 Name: ps_ind_09_bin, dtype: int64, 0    594990
 1       222
 Name: ps_ind_10_bin, dtype: int64, 0    594205
 1      1007
 Name: ps_ind_11_bin, dtype: int64, 0    589594
 1      5618
 Name: ps_ind_12_bin, dtype: int64, 0    594648
 1       564
 Name: ps_ind_13_bin, dtype: int64, 1    393330
 0    201882
 Name: ps_ind_16_bin, dtype: int64, 0    523143
 1     72069
 Name: ps_ind_17_bin, dtype: int64, 0    503879
 1     91333
 Name: ps_ind_18_bin, dtype: int64,

The table above is the result fo Q2, and since the format is not well done, please read it carefully (some are horizontal and some are in the next line).

## Q3: How would you compute similarity between feature vectors of any pair of samples (rows)? Please first describe the steps or formula you will use to compute similarities. We require that the similarity measure should have values between 0 and 1. Then implement your function in script that takes two row indices as input and return a similarity value as output. (We require that your function  should be within 20 lines of codes)!

### Answer 
##### For numerical data:

Steps to compute similarities

1. Compute cosine similarity between samples in X and Y
2. Cosine similarity, or the cosine kernel, computes similarity as the
    normalized dot product of X and Y:
        K(X, Y) = <X, Y> / (||X||*||Y||)
3. Parameters

    X : ndarray or sparse array, shape: (n_samples_X, n_features)
        Input data.
    Y : ndarray or sparse array, shape: (n_samples_Y, n_features)
        Input data. If ``None``, the output will be the pairwise
        similarities between all samples in ``X``.
    dense_output : boolean (optional), default True
        Whether to return dense output even when the input is sparse. If
        ``False``, the output is sparse if both input arrays are sparse.


##### For categorical data:
We just compare the values for the categorical data to decide whether they are the same or not.

##### Some pre-analysis for the dataset before the function
use the package as well as function to prove the answer

In [18]:
row_range_1 = 3
row_range_2 = 5

In [19]:
df[row_range_1:row_range_2]  ## selected data (two rows)

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
3  16       0          0              1          2              0   
4  17       0          0              2          0              1   

   ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin       ...        \
3              0              1              0              0       ...         
4              0              1              0              0       ...         

   ps_calc_11  ps_calc_12  ps_calc_13  ps_calc_14  ps_calc_15_bin  \
3           2           2           4           9               0   
4           3           1           1           3               0   

   ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  ps_calc_19_bin  \
3               0               0               0               0   
4               0               0               1               1   

   ps_calc_20_bin  
3               0  
4               0  

[2 rows x 59 columns]

In [22]:
## first use cosine similarity package to calculate the result
from sklearn.metrics.pairwise import cosine_similarity   
cosine_similarity(df[row_range_1:row_range_2])

array([[ 1.        ,  0.98359439],
       [ 0.98359439,  1.        ]])

In [99]:
## Then use the function to calculate the similarity
from scipy.spatial.distance import cosine   

data = df[col_Q3_1].loc[[row_range_1,row_range_2]].values
m, k = data.shape

mat = np.zeros((m, m))

for i in range(m):
    for j in range(m):
        if i != j:
            mat[i][j] = 1- cosine(data[i,:], data[j,:])
        else:
            mat[i][j] = 1.

In [100]:
mat  ## check if it is the same as the cosine similarity result

array([[ 1.        ,  0.92067215],
       [ 0.92067215,  1.        ]])

### Answer (function): The function to compare the similarities (20 lines in tatal without notations)
steps:
1. For numerical data, I will use cosine similarities
2. For categorical data, I will use 1 to represent the same value, 0 to represent difference.
3. Then calculate the mean of the categorical similarity results (to normalize the result), as simi_cat
4. Combine the numerical and categorical value by using their weight, for example there are 7 columns are numerical and 5 are categorical, then calculate the weight average as the final similarity result


In [139]:
from scipy.spatial.distance import cosine   

df_Q3 = df.drop(['id'],axis=1)
col_Q3 = df_Q3.columns[[v[-3:] in 'cat' for v in df_Q3.columns]]  ## columns that are categorical
col_Q3_num = [vv for vv in list(df_Q3.columns) if vv not in list(col_Q3)]  ## columns that are numerical
## or we can use
## col_Q3_num = set(list(df_Q3.columns)) - set(list(col_Q3))
def compare_similarities(row_range_1,row_range_2,df_new):
    # For categorical data
    ## num_zero_cat is the number of zero in each row, which means the number of value that equals in two rows
    ## num_non_zero_cat is the number of columns that catgorical data are differnt in two selected rows
    num_zero_cat = (df_new[col_Q3].loc[[row_range_1, row_range_2]].diff().iloc[[1]] == 0).astype(int).sum(axis=1).sum()  
    num_non_zero_cat = (len(col_Q3) - num_zero_cat)*1
    simi_cat = (num_zero_cat*0 + num_non_zero_cat*1)/len(col_Q3)  ## normalized the similarities of categorical data
    # For numerical data
    data_num = df_new[col_Q3_num].loc[[row_range_1, row_range_2]].values
    m1,m2 = data_num.shape
    mat_num = np.zeros((m1,m1))
    for i in range(m1):
        for j in range(m1):
            if i != j:
                mat_num[i][j] = 1 - cosine(data_num[i,:], data_num[j,:])
            else:
                mat_num[i][j] = 1
    ## extract the cosine similarity from the matrix
    simi_num = mat_num[0,1]
    # Combine the categorical similarities together with numerical similarities
    similarity_result = (len(col_Q3)/len(df_new.columns))*simi_cat + (len(col_Q3_num)/len(df_new.columns))*simi_num
    return similarity_result

In [140]:
compare_similarities(row_range_1,row_range_2,df_Q3)

0.756443697690665

## Q4: How many features contain missing values? For each feature containing missing values, what is the ratio of rows (samples) that miss values? How many samples in total contain missing values? 

### Answer 
1. Missing value: NaN(none of the feature); -1 (13 features)
2. The ratio for each feature are as below:

feature name:ps_ind_02_cat,  # of missing value:216,  ratio of rows: 0.0003628959093566662

feature name:ps_ind_04_cat,  # of missing value:83,  ratio of rows: 0.00013944611331760784

feature name:ps_ind_05_cat,  # of missing value:5809,  ratio of rows: 0.00975954785857812

feature name:ps_reg_03,  # of missing value:107772,  ratio of rows: 0.18106489788512328

feature name:ps_car_01_cat,  # of missing value:107,  ratio of rows: 0.0001797678810239041

feature name:ps_car_02_cat,  # of missing value:5,  ratio of rows: 8.400368272145051e-06

feature name:ps_car_03_cat,  # of missing value:411231,  ratio of rows: 0.6908983689844963

feature name:ps_car_05_cat,  # of missing value:266551,  ratio of rows: 0.4478253126617071

feature name:ps_car_07_cat,  # of missing value:11489,  ratio of rows: 0.019302366215734897

feature name:ps_car_09_cat,  # of missing value:569,  ratio of rows: 0.0009559619093701067

feature name:ps_car_11,  # of missing value:5,  ratio of rows: 8.400368272145051e-06

feature name:ps_car_12,  # of missing value:1,  ratio of rows: 1.68007365442901e-06

feature name:ps_car_14,  # of missing value:42620,  ratio of rows: 0.07160473915176441

In [148]:
## first, check if there is NAN in the dataset
df.isnull().any() 

id                False
target            False
ps_ind_01         False
ps_ind_02_cat     False
ps_ind_03         False
ps_ind_04_cat     False
ps_ind_05_cat     False
ps_ind_06_bin     False
ps_ind_07_bin     False
ps_ind_08_bin     False
ps_ind_09_bin     False
ps_ind_10_bin     False
ps_ind_11_bin     False
ps_ind_12_bin     False
ps_ind_13_bin     False
ps_ind_14         False
ps_ind_15         False
ps_ind_16_bin     False
ps_ind_17_bin     False
ps_ind_18_bin     False
ps_reg_01         False
ps_reg_02         False
ps_reg_03         False
ps_car_01_cat     False
ps_car_02_cat     False
ps_car_03_cat     False
ps_car_04_cat     False
ps_car_05_cat     False
ps_car_06_cat     False
ps_car_07_cat     False
ps_car_08_cat     False
ps_car_09_cat     False
ps_car_10_cat     False
ps_car_11_cat     False
ps_car_11         False
ps_car_12         False
ps_car_13         False
ps_car_14         False
ps_car_15         False
ps_calc_01        False
ps_calc_02        False
ps_calc_03      

In [149]:
df.isnull().any().sum()  ## which means that none of the feature contains NAN

0

In [150]:
## Then, we can see that there are some negative value in the dataset
## However, it is not possible that '-1' shows up here, so it can also be seen as missing value (total 13 columns here)
df[(df == -1)].any().sum()

13

In [151]:
df.columns[df[(df == -1)].any()]  ## this is the columns which has '-1' as missing value

Index(['ps_ind_02_cat', 'ps_ind_04_cat', 'ps_ind_05_cat', 'ps_reg_03',
       'ps_car_01_cat', 'ps_car_02_cat', 'ps_car_03_cat', 'ps_car_05_cat',
       'ps_car_07_cat', 'ps_car_09_cat', 'ps_car_11', 'ps_car_12',
       'ps_car_14'],
      dtype='object')

In [124]:
col_Q4 = list(df.columns[df[(df == -1)].any()])

In [161]:
for c in col_Q4:
    print('feature name:%s, '%c,
          '# of missing value:%s, '%((df[c])<0).sum(),
          'ratio of rows: %s' %((((df[c])<0).sum()).astype(int)/len(df[c])))

feature name:ps_ind_02_cat,  # of missing value:216,  ratio of rows: 0.0003628959093566662
feature name:ps_ind_04_cat,  # of missing value:83,  ratio of rows: 0.00013944611331760784
feature name:ps_ind_05_cat,  # of missing value:5809,  ratio of rows: 0.00975954785857812
feature name:ps_reg_03,  # of missing value:107772,  ratio of rows: 0.18106489788512328
feature name:ps_car_01_cat,  # of missing value:107,  ratio of rows: 0.0001797678810239041
feature name:ps_car_02_cat,  # of missing value:5,  ratio of rows: 8.400368272145051e-06
feature name:ps_car_03_cat,  # of missing value:411231,  ratio of rows: 0.6908983689844963
feature name:ps_car_05_cat,  # of missing value:266551,  ratio of rows: 0.4478253126617071
feature name:ps_car_07_cat,  # of missing value:11489,  ratio of rows: 0.019302366215734897
feature name:ps_car_09_cat,  # of missing value:569,  ratio of rows: 0.0009559619093701067
feature name:ps_car_11,  # of missing value:5,  ratio of rows: 8.400368272145051e-06
feature na

In [168]:
(df[df<0].count(axis=1)).value_counts()

2    254982
1    156295
0    124931
3    56042 
4    2883  
7    53    
6    19    
8    7     
dtype: int64

In [167]:
## How many samples in total contain missing values? --- 470281
(df[df<0].count(axis=1)>0).value_counts()

True     470281
False    124931
dtype: int64

## Q5:	Please fill in the missing values, and briefly describe your approach.

### Answer

If we want to fill in the missing value, there are several different cases needed to be considered:
1. In previous Q4, we can see that if the ratio of missing value > 20%, I decided to drop this column
Then
2. For numerical data, we can calculate the mean/median value of this column, then use this value to impute the missing place
3. For categorical data, since we need to do the one-hot encoding afterwards, we can keep the -1 here as a specific class (used for later dummies data, which means this spot is missing)


In [188]:
col_Q5 = []
df_impute = df.copy()
for c in col_Q4:
    if ((((df[c])<0).sum()).astype(int)/len(df[c])) < 0.2:
        col_Q5.append(c)
        if c[-3:] != 'cat':
            print(c)
            df_impute[c] = np.where(df[c]==-1, df_impute[c].median(), df_impute[c] )

ps_reg_03
ps_car_11
ps_car_12
ps_car_14


In [170]:
col_Q5

['ps_ind_02_cat',
 'ps_ind_04_cat',
 'ps_ind_05_cat',
 'ps_reg_03',
 'ps_car_01_cat',
 'ps_car_02_cat',
 'ps_car_07_cat',
 'ps_car_09_cat',
 'ps_car_11',
 'ps_car_12',
 'ps_car_14']

## Q6: How many classes are there in our target column? Is our class balanced or highly imbalanced? What challenge do you expect in classification task based on your observation?

### Answer
1. There are two classes in our target column: label '0': 573518, label '1': 21694. 

2. Therefore, the class is imbalanced, label '1' : label '0' =  0.0378. 

3. In this case, the label is imbalanced. When we do the classfication, sometimes the 'accuracy' metric will be very high. However, it is not true since maybe most of the label '0' are predicted correctly, while label '1' has very low accuracy. Therefore, at this time, accuracy cannot be the exact metric to evaluate the performance and it will mislead the model performance.

In [190]:
df_impute.columns

Index(['id', 'target', 'ps_ind_01', 'ps_ind_02_cat', 'ps_ind_03',
       'ps_ind_04_cat', 'ps_ind_05_cat', 'ps_ind_06_bin', 'ps_ind_07_bin',
       'ps_ind_08_bin', 'ps_ind_09_bin', 'ps_ind_10_bin', 'ps_ind_11_bin',
       'ps_ind_12_bin', 'ps_ind_13_bin', 'ps_ind_14', 'ps_ind_15',
       'ps_ind_16_bin', 'ps_ind_17_bin', 'ps_ind_18_bin', 'ps_reg_01',
       'ps_reg_02', 'ps_reg_03', 'ps_car_01_cat', 'ps_car_02_cat',
       'ps_car_03_cat', 'ps_car_04_cat', 'ps_car_05_cat', 'ps_car_06_cat',
       'ps_car_07_cat', 'ps_car_08_cat', 'ps_car_09_cat', 'ps_car_10_cat',
       'ps_car_11_cat', 'ps_car_11', 'ps_car_12', 'ps_car_13', 'ps_car_14',
       'ps_car_15', 'ps_calc_01', 'ps_calc_02', 'ps_calc_03', 'ps_calc_04',
       'ps_calc_05', 'ps_calc_06', 'ps_calc_07', 'ps_calc_08', 'ps_calc_09',
       'ps_calc_10', 'ps_calc_11', 'ps_calc_12', 'ps_calc_13', 'ps_calc_14',
       'ps_calc_15_bin', 'ps_calc_16_bin', 'ps_calc_17_bin', 'ps_calc_18_bin',
       'ps_calc_19_bin', 'ps_calc_20_bin'],


In [191]:
df_impute.shape

(595212, 59)

In [203]:
df_impute.target.value_counts()

0    573518
1    21694 
Name: target, dtype: int64

In [204]:
(df_impute['target']==1).sum()/(df_impute['target']==0).sum()  ## the ratio of imbalanced data (1:0)

0.037826188541597645

## Q7:	Suppose we need to reduce the feature dimension to m (m is a parameter such as 10), and decided to use Principle Component Analysis (PCA) to do that. Can you directly run PCA on our data and why? 

If not, please preprocess the data and then run PCA with m = 10. In your results, please provide the 10 principle components (vectors) by a decreasing order, as well as data with reduced dimension.

Hint: for categorical attributes, we can encode them into a set of binary attributes (see https://www.kdnuggets.com/2015/12/beyond-one-hot-exploration-categorical-variables.html). 

### Answer 
1. No, it cannot directly use PCA since there are some categorical data in the dataframe, since PCA can only deal with the linear relationship numerical data. Therefore, we should use one-hot encoding to first transform categorical data into numerical data, then use PCA to reduce features.

2. Also, we still need to standardize the values. PCA is effected by scale so we need to scale the features in the data before applying PCA. (https://towardsdatascience.com/pca-using-python-scikit-learn-e653f8989e60)

In [207]:
col_Q7 = df_impute.columns[[v[-3:] in 'cat' for v in df_impute.columns]]

In [208]:
col_Q7

Index(['ps_ind_02_cat', 'ps_ind_04_cat', 'ps_ind_05_cat', 'ps_car_01_cat',
       'ps_car_02_cat', 'ps_car_03_cat', 'ps_car_04_cat', 'ps_car_05_cat',
       'ps_car_06_cat', 'ps_car_07_cat', 'ps_car_08_cat', 'ps_car_09_cat',
       'ps_car_10_cat', 'ps_car_11_cat'],
      dtype='object')

In [219]:
df_new = pd.get_dummies(df_impute, columns=col_Q7)

In [223]:
df_impute.shape

(595212, 59)

In [222]:
df_new.shape

(595212, 229)

In [224]:
df_new.columns

Index(['id', 'target', 'ps_ind_01', 'ps_ind_03', 'ps_ind_06_bin',
       'ps_ind_07_bin', 'ps_ind_08_bin', 'ps_ind_09_bin', 'ps_ind_10_bin',
       'ps_ind_11_bin',
       ...
       'ps_car_11_cat_95', 'ps_car_11_cat_96', 'ps_car_11_cat_97',
       'ps_car_11_cat_98', 'ps_car_11_cat_99', 'ps_car_11_cat_100',
       'ps_car_11_cat_101', 'ps_car_11_cat_102', 'ps_car_11_cat_103',
       'ps_car_11_cat_104'],
      dtype='object', length=229)

In [229]:
col_x = df_new.columns[2:]

In [232]:
from sklearn.preprocessing import StandardScaler

features = col_x
# Separating out the features
x = df_new.loc[:, features].values
# Separating out the target
y = df_new.loc[:,['target']].values
# Standardizing the features
x = StandardScaler().fit_transform(x)

In [246]:
from sklearn.decomposition import PCA
pca = PCA(n_components=10)
principalComponents = pca.fit_transform(x)
principalDf = pd.DataFrame(data = principalComponents
             , columns = ['principal_component_{0}'.format(i) for i in range(10)])

In [248]:
principalDf

principal_component_0  principal_component_1  principal_component_2  \
0       0.728200               0.172518              -0.881308                
1      -1.828792              -0.611185              -0.254239                
2      -2.377871              -0.364436              -0.756170                
3      -1.899396               0.748594               2.103025                
4      -2.571529               0.588896              -2.801840                
5       3.466936              -2.451428              -1.752773                
6      -1.496763              -1.898499               0.525457                
7      -0.001749               0.782599               0.152227                
8       0.233281              -2.554005              -0.975313                
9       6.168378               1.389807              -0.562906                
10      3.163615              -2.445844              -0.419530                
11      1.439971               0.426515              -1.738819                
12     -1.505563              -1.052972              -0.981399                
13      1.283613               1.885801              -0.503522                
14     -1.711732              -0.567142              -0.350036                
15      6.995928              -4.102358               1.696512                
16      0.705163               0.373901              -0.781527                
17      0.445339               3.679966               0.921182                
18     -2.287771              -0.919639              -0.351709                
19     -3.889571              -0.235392              -0.826889                
20     -2.001995               0.046240              -1.104454                
21     -2.946199               0.539969              -1.150041                
22      2.145453               0.921438              -0.567669                
23     -0.689680              -2.151683               0.642013                
24     -0.702051              -2.032439               1.970377                
25      1.860742              -5.466269               3.203247                
26      2.314336              -1.760110               0.019337                
27     -1.223108              -1.983099               0.650127                
28      0.971993              -2.170480              -0.910064                
29      1.112678              -2.073157              -2.697752                
...          ...                    ...                    ...                
595182 -1.421781              -1.573541               2.705621                
595183 -2.607170              -0.000988              -0.289655                
595184 -0.932200              -2.261062               6.599238                
595185 -1.337283               4.228086               0.429800                
595186 -0.016348               4.111154              -0.563171                
595187  1.024977              -3.035452               4.583749                
595188 -2.138834              -0.194570              -0.208758                
595189 -1.401252               0.191129               1.520855                
595190 -3.041662              -1.926421               1.380942                
595191 -0.394881              -0.820626              -0.218962                
595192 -1.242144               1.324328              -0.964088                
595193  3.989743               5.890997               1.741228                
595194 -1.230752               1.525347               1.236634                
595195  2.114903               3.538532               0.551044                
595196 -0.455136              -2.015755               5.918610                
595197  2.616276               0.920938              -1.267856                
595198  0.049014               1.951480              -1.001671                
595199  1.197010               3.085651               0.244844                
595200 -2.195547               0.135047              -0.41